### ISSUE TITLE
Ability to use custom check attributes in command execution

### ISSUE THREAD

**Issue by mvdavid7** [sensu/sensu#1586]
I've tried asking about this in irc and the mailing list, but haven't gotten a definitive answer. 

The documentation for checks says that "Custom check definition attributes may be defined to provide context about the check, or to control or modify check command or handler behaviors." (https://sensuapp.org/docs/latest/reference/checks.html#custom-attributes)

This implies to me that i could set custom attributes in my check definition and then use them from the command. How do I do that? Command token substitution seems to apply only to custom data in the client config.

## Expected Behavior
I should be able to use custom check attributes in a check command.

## Current Behavior
Only client custom attributes are exposed via command token substitution

## Context
I have various different environments (QA, staging, production) being monitored by the same sensu server. We have checks that apply to all environments, but need different arguments (like a website status check that monitors if the website is up on each of these environments). I know i can use the same check by using command token substitution and defining the url of the particular site to test in the client config (i.e.: prod.example.com on the prod client, and qa.example.com on the qa client).

However, this is less than ideal because that means for any check that needs to be parameterized, i need to modify every client that will run the check (which even with automated infrastructure, i'd still rather avoid doing work on prod machines).

Ideally, i would be able to define all the environment specific parameters in the check definition and just pull one parameter from the client, the environment name. Then when the check executes, it would use the environment name to select the appropriate parameter for the rest of the check options.

**Comment by cwjohnston** [comment]
Hi @mvdavid7,

> This implies to me that i could set custom attributes in my check definition and then use them from the command. How do I do that? Command token substitution seems to apply only to custom data in the client config.

I can see what you mean, the documentation is definitely wrong here. The current reality is that custom check attributes are not available for use in token substitution. I opened https://github.com/sensu/sensu-docs/issues/519 to track the need to address this.

> I should be able to use custom check attributes in a check command.

I suggest that you check out the proxy check requests feature we've just added in Sensu 0.28. Documentation for this new feature is sparse as we just shipped the release yesterday, but, if I understand your use case, I think it may help.

In a nutshell:

1. Create proxy clients (e.g. representations of the websites you need to check) with appropriate attributes

2. Create or configure a check definition with [`proxy_requests` attributes](https://sensuapp.org/docs/latest/reference/checks.html#proxy-requests-attributes) to match these clients. 

Once scheduled, the single check definition will generate a check request for each client matching the specified proxy_requests values, and perform token substitution using each matching client's attributes.

See https://github.com/sensu/sensu/pull/1555#issuecomment-280394697 for another user's interpretation of how this is intended to work.

**Comment by calebhailey** [comment]
@mvdavid7 hi there :wave: 

That bit of documentation is a bit misleading (and I'm pretty sure I wrote it, so I'm sorry about that!)... custom attributes _can_ be used to do exactly what you're describing, however **not** via [token substitution](https://sensuapp.org/docs/latest/reference/checks.html#check-token-substitution). 

Token substitution (via custom _client_ attributes) may be the best solution for what you're describing, although it seems you'd rather not pursue that path. The alternative, as the documentation very opaquely eludes to, is that a custom monitoring plugin itself could leverage conditional logic using the custom check definition attributions (e.g. "if `client.environment` is `production`, use `check.custom_attribute`"). 

I hope this helps!

**Comment by cwjohnston** [comment]
I think I followed up with Caleb about this out of band some time ago, so let me add some thoughts about his response here.

> Token substitution (via custom _client_ attributes) may be the best solution for what you're describing, although it seems you'd rather not pursue that path. The alternative, as the documentation very opaquely eludes to, is that a custom monitoring plugin itself could leverage conditional logic using the custom check definition attributions (e.g. "if `client.environment` is `production`, use `check.custom_attribute`"). 

The Sensu check plugin specification (identical to that of Nagios) does not allow for check plugins to receive check request details, e.g. via STDIN. As a result, a check plugins lack the context of the execution request that spawns them, aside from the environment and flags specified in the `command` field. The only way a custom check plugin may access this context to evaluate custom check attributes would be to load relevant check configuration from disk. This may be possible but likely not feasible in every case.
